File: kotlin/src/main/kotlin/leetcode/887.kt
package leetcode

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class SuperEggDrop887 {
    fun superEggDrop(k: Int, n: Int): Int {
        val max_floors = MutableList(n + 1) { 0 }
        var res = 0
        repeat(k) {
            var pre = 0
            for (t in 1..n) {
                max_floors[t] = (1 + max_floors[t - 1] + pre).also {
                    pre = max_floors[t]
                }
                if(max_floors[t] >= n){
                    res = t
                    break
                }
            }
        }
        return res
    }

    fun superEggDrop1(k: Int, n: Int): Int {
        val min = List(2) { MutableList(n + 1) { it } }
        for (i in 2..k) {
            var cur = min[i and 1]
            var last = min[i and 1 xor 1]
            var t = 1
            for (x in 2..n) {
                while (t < x && cur[t] < last[x - 1 - t]) t += 1
                cur[x] = 1 + cur[t]
            }
        }
        return min[k and 1][n]
    }


    @Test
    fun test1() {
        assertEquals(2, superEggDrop(1, 2))
        assertEquals(3, superEggDrop(2, 6))
        assertEquals(4, superEggDrop(3, 14))
    }
}

File: golang/leetcode/887.go
package main

func superEggDrop1(k int, n int) int {
	dp := make([]int, n+1)
	for i := range dp {
		dp[i] = i
	}
	for i := 2; i <= k; i++ {
		ndp := make([]int, n+1)
		ndp[1] = 1
		for j := 2; j <= n; j++ {
			l, r := 1, j
			for l <= r {
				md := (l + r) >> 1
				if ndp[j-md] <= dp[md-1] {
					r = md - 1
				} else {
					l = md + 1
				}
			}
			ndp[j] = 1 + max(dp[l-1], ndp[j-l])
		}
		dp = ndp
	}
	return dp[n]
}

func superEggDrop(k int, n int) int {
	if n == 1 {
		return 1
	}
	dp := make([]int, k+1)
	for i := 1; i <= k; i++ {
		dp[i] = 1
	}
	for t := 2; ; t++ {
		for i := k; i >= 1; i-- {
			dp[i] += 1 + dp[i-1]
		}
		if dp[k] >= n {
			return t
		}
	}
}
